### ISSUE TITLE
Ways to improve backend performance

### ISSUE THREAD

**Issue by ghost** [octobercms/october#4379]
I'm finding October CMS runs slow in the backend when loading between web pages and would like to open a discussion about this.

Having a general think about things, we have many files that are not minified (for testing reasons).

The files I find that are slowing things down are the main javascript files in the backend things such as: `storm-min.js` etc.

p.s. `storm-min.js`, `october-min.js` and `october.flyout.js` seem to be the worst performing files by quite a margin over the rest.

`october.flyout.js` slow issue is related to an open issue found here: https://github.com/octobercms/october/issues/4298 and https://github.com/octobercms/october/issues/4052

Other factors that make October feel slow is that I'm using CloudFlare which I have seen a few times: Railgun Errors.

Another factor maybe the large number of plugins I have running.

## Speed Loading Guide

Below is a table showing the speed performance guide by Google on how fast a web page should load, note the backend web pages should in theory loaded under 100mS.

![small-delay-383x300](https://user-images.githubusercontent.com/17784082/59203940-85a4bf80-8b97-11e9-94e6-cf0950920f68.png)

As developers we aim for '**instant loading**'. 

p.s. My front-end loads instantly, yet the backend is slow and sluggish.

## Service Worker

One solution I would like to raise and discuss is to make some of these files offline and cached.

By using a Service Worker in the Backend.

I know this would cause an issue when October CMS updates and to get around that issue I would suggest adding a Service Worker version number the same as the current October CMS. By making this variable dynamic and auto updating when October CMS is updated then the Service Worker would clear the cache when October CMS is updated - thus giving those files a refresh.

e.g.

`storm-min.js?v=456`

would be:

```javascript
workbox.core.setCacheNameDetails({
    suffix: '<?= $script . \'?v=\' . $coreBuild; ?>',
});
```

I would suggest making the main JavaScript and CSS files cached and offline.

## "304 Not Modified" response

Another idea, would be to using a caching method to speed up the loading of the same web page give a 304 Service Code instead of another 200 Service Code.

## Final thoughts

These things should speed up October CMS loading times and not rely so heavily on the Network Connection (not giving any issues). 

I leave it open to people's thoughts and ideas.

**Comment by ayumihamsaki** [comment]
p.s. Please commit below is anyone sees any performance issues with the backend, would be interested to know how many people are affected by this related issue?

**Comment by alxy** [comment]
When you are talking about loading times/performance/speed - are you talking about the time it takes for the server to provide the requested files? (Which will depend on their respective sizes and your internet connection.) The specific asset files you mention are typically cached in the browser, so they should only significantly affect performance on the first load. And this is what happens for my october installations as well: basically all backen assets are loaded from disk, which is pretty fast.

**Comment by LukeTowers** [comment]
@ayumihamsaki I would say the best place to start is by profiling the backend to see where the biggest performance slowdowns are and work on minimizing them one by one with specific PRs.

**Comment by FlusherDock1** [comment]
I don't have any performance issues in the local or production environment. How many plugins do you have? How many of them have their additional assets?

The idea of ​​using ServiceWorker is very good. I think it is possible to implement it also to the frontend core framework.

**Comment by ayumihamsaki** [comment]
Will add some more info:

1. I got around 60 plugins running, I did open dev-tools to check to see if any of the assets are slowing down the loading of the web pages, but they seem to not be the issue.

2. ServiceWorker on the front-end, I'm not sure if that would be a good idea, due to everyone having different requirements, I do have a Service Worker running on the front-end. But I would say to get it working right you really need to fine tune it and customize it to the code to run properly. I don't think you could get a generalized Service Worker to work perfectly for all October websites. I know Luke would be screaming "You can do that in a plugin" lol.

3. Any issues and findings I will open up other github issues and create specific PRs.

4. It seems like there is a large "**_long-task_**" which looks like it's causing the delay by around 3000mS, please see screenshot:

![Untitled](https://user-images.githubusercontent.com/17784082/59224292-f06cef80-8bc5-11e9-97fc-34164ec7fc8e.png)

Lastly, I'm using Google Canary, so that may also be causing some issues! As I have the experiments tag open and I can see Google working on the new the API's / HTTP Headers:

```
sec-ch-ua: Google Chrome 77
Sec-Fetch-Dest: document
Sec-Fetch-Mode: navigate
Sec-Fetch-User: ?1
Sec-Origin-Policy: 0
```

Can see these security headers being added and as they are beta that may also be causing an issue.

I would expect opening October in Firefox or Edge it would be quicker my side.

But I really think the **long-tasks** are slowing done the web page from loading and I know code splitting would help in that case for the javascript.

**Comment by ayumihamsaki** [comment]
Will add some more info:

1. I got around 60 plugins running, I did open dev-tools to check to see if any of the assets are slowing down the loading of the web pages, but they seem to not be the issue.

2. ServiceWorker on the front-end, I'm not sure if that would be a good idea, due to everyone having different requirements, I do have a Service Worker running on the front-end. But I would say to get it working right you really need to fine tune it and customize it to the code to run properly. I don't think you could get a generalized Service Worker to work perfectly for all October websites. I know Luke would be screaming "You can do that in a plugin" lol.

3. Any issues and findings I will open up other github issues and create specific PRs.

4. It seems like there is a large "**_long-task_**" which looks like it's causing the delay by around 338mS, please see screenshot:

![1](https://user-images.githubusercontent.com/17784082/59225123-d3d1b700-8bc7-11e9-943e-10517d52344e.png)

5. Also I'm seeing a long Idle Time: 1650mS which could also be reduced with an API, see here:

![2](https://user-images.githubusercontent.com/17784082/59225124-d3d1b700-8bc7-11e9-8fbe-26900b58b9cc.png)

Lastly, I'm using Google Canary, so that may also be causing some issues! As I have the **experiments tag** open and I can see Google working on the new the API's / HTTP Headers:

```
sec-ch-ua: Google Chrome 77
Sec-Fetch-Dest: document
Sec-Fetch-Mode: navigate
Sec-Fetch-User: ?1
Sec-Origin-Policy: 0
```

Can see these new security headers being added to the web page and they are in beta so they maybe slowing down the rendering process?

I would expect opening October in Firefox or Edge would be quicker my side.

Adding a Service Worker and some performance API's (for the Idle Load) and code splitting (for the long-tasks) may fix some of the issues.

Idle Load API:

https://philipwalton.com/articles/idle-until-urgent/
https://developers.google.com/web/updates/2015/08/using-requestidlecallback
https://gist.github.com/paullewis/55efe5d6f05434a96c36

Long Tasks with Code Splitting:

https://developers.google.com/web/fundamentals/performance/optimizing-javascript/code-splitting/
https://w3c.github.io/longtasks/

**Comment by ayumihamsaki** [comment]
Gonna write this comment quickly, my gut feeling is telling me that the October backend has a bad render blocking issue as well.

Currently it is getting loaded from assets like `storm-min.js` and if that gets '**stalled**' or '**delayed**' etc. then the rendering of the web page gets delayed.

In fact Google Chrome V77 is telling me `storm-min.js` is getting '**stalled**'.

A better way to do the rendering would be to use the App Shell model, see here: https://developers.google.com/web/fundamentals/architecture/app-shell

In this method we load the backend like this:

![1](https://user-images.githubusercontent.com/17784082/59228492-dc2df000-8bcf-11e9-95fb-68468b596cdf.png)

Done by inline CSS and then fill in the things with the assets async way.

We would fill in each thing one at a time and not all at the same time! Which is going on right now.

For example:

1. The shell loads
2. The top menu loads next
3. The side bars load after that
4. The main content loads last

If part 2 (the top menu) gets stuck e.g. `october.flyout.js` the rest of the web page will load without that part.

This way we don't start with a blank screen and try to load everything at the same time and have to wait several seconds starring at a white screen.

**Comment by ayumihamsaki** [comment]
Gonna write this comment quickly, my gut feeling is telling me that the October backend has a bad render blocking issue as well.

Currently it is getting loaded from assets like `storm-min.js` and if that gets '**stalled**' or '**delayed**' etc. then the rendering of the web page gets delayed.

In fact Google Chrome V77 is telling me `storm-min.js` is getting '**stalled**'.

A better way to do the rendering would be to use the App Shell model, see here: https://developers.google.com/web/fundamentals/architecture/app-shell

In this method we load the backend like this:

![1](https://user-images.githubusercontent.com/17784082/59228492-dc2df000-8bcf-11e9-95fb-68468b596cdf.png)

Done by inline CSS and then fill in the things with the assets async way.

We would fill in each thing one at a time and not all at the same time! Which is going on right now.

For example:

1. The shell loads
2. The top menu loads next
3. The side bars load after that
4. The main content loads last

If part 2 (the top menu) gets stuck e.g. `october.flyout.js` the rest of the web page will continue loading and not get stuck by that part.

This way we don't start with a blank screen and try to load everything at the same time and have to wait several seconds starring at a white screen.

Here are three methods showing you some examples:

1. Normal Loading (Old Way):

![1](https://user-images.githubusercontent.com/17784082/59229215-b7d31300-8bd1-11e9-827d-73a4505084ee.gif)

2. Filling the screen (New Way):

![2](https://user-images.githubusercontent.com/17784082/59229238-c8838900-8bd1-11e9-83cd-bcd879e590dc.gif)

3. Staggering (New Way):

![3](https://user-images.githubusercontent.com/17784082/59229275-da652c00-8bd1-11e9-9e40-e82f3711d881.gif)

**Comment by FlusherDock1** [comment]
> Lastly, I'm using Google Canary, so that may also be causing some issues!

I just tested some of my biggest projects, with a lot of records and plugins on Chrome Stable, and I always have around 50ms on scripting task.

Can you please test it on other browsers as well?

**Comment by ayumihamsaki** [comment]
Ignore my comment about the browser, because it wouldn't really matter what browser you use with October. You will still get the same issues, such as: rendering blocking, everything trying to be loaded all at once and network congestion issues.

The only way to fix this would be to re-code October using modern coding practises like the [App Shell Model](https://developers.google.com/web/fundamentals/architecture/app-shell), [R.A.I.L.](https://developers.google.com/web/fundamentals/performance/rail) and [PRPL Pattern](https://developers.google.com/web/fundamentals/performance/prpl-pattern/).

Also maybe having different style code into master and develop. **Master** to have minifed versions for speed and performance and then **Develop** to have un-minifed versions for testing.

The issue I have right now is two things:

1. Rendering Blocking - (trying to load everything at once and stalling with something having a bug)
2. Network Congestion - (October replies on the internet and can not work offline)

**Comment by ayumihamsaki** [comment]
Ignore my comment about the browser, because it wouldn't really matter what browser you use with October. You will still get the same issues, such as: rendering blocking, everything trying to be loaded all at once and network congestion issues.

The only way to fix this would be to re-code October using modern coding practises like the [App Shell Model](https://developers.google.com/web/fundamentals/architecture/app-shell), [R.A.I.L.](https://developers.google.com/web/fundamentals/performance/rail) and [PRPL Pattern](https://developers.google.com/web/fundamentals/performance/prpl-pattern/).

Also maybe having different style code into master and develop. **Master** to have minifed versions for speed and performance and then **Develop** to have un-minifed versions for testing.

The issue I have right now is two things:

1. Rendering Blocking - (trying to load everything at once and stalling with something having a bug)
2. Network Congestion - (October replies on the internet and can not work offline)

A third minor thing would also be there are no **304 redirects** setup to reduce the resources. 

Lastly, I have a feeling October probably fails the number of **DOM elements** and **tree depth**. It would not surprise me if October uses far too many of these.

**Comment by LukeTowers** [comment]
@ayumihamsaki it's highly unlikely that October is going to be rewritten in any major way, especially for the sake of checking of boxes of random online checklists / design patterns.

We are more than happy to deal with individual performance issues as they come up and are discussed in detail but I'm not seeing anyone else having the same problems as you. If you can make PRs for individual changes that you think will improve the performance along with a detailed background of the change and performance profiling for before and after (with server specs, browser details, and active load) then go for it.

I know there are definitely some individual pieces of the october backend where performance could be improved (for instance one of my projects heavily utilizes the dashboard with lots of custom dashboard widgets and the performance of it loading is abysmal), but everything should be addressed on an individual basis with detailed profiling of the performance before / after the change to be sure we're making changes that actually improve October and not just changes for changes sake.

**Comment by Samuell1** [comment]
First things that should be optimalized is to recompile all assets and minify them trough something better for example "framework.extras-min.css" can save a lot in size.

(first is custom recompiled trough postcss)
![obrázok](https://user-images.githubusercontent.com/3110002/59332934-a2580900-8cf7-11e9-93c6-0284a814bc17.png)

**Comment by bennothommo** [comment]
@ayumihamsaki Regarding ""304 Not Modified" response - that's something that is usually controlled at the server level. The only way we can enforce that is by routing asset calls through the PHP pipeline, which would result in even worse loading times initially. We could possibly add a doc page suggesting best practices for hosting, of which this could be one of them.

**Comment by ayumihamsaki** [comment]
@bennothommo I wrote my own blog plugin for October, due to two reasons why I didn't like RainLab's Blog plugin and RadiantWeb's ProBlog Plugin. The two issues I found was:

1. They both have duplicate content issues creating web pages in their folder structures.
2. They both didn't have 304 Redirects.

In my Blog Plugin I did use php as you suggested to create the 304 redirect, the code example I used was a modified version found here: https://www.php.net/manual/en/function.header.php

I have not seen any performance issues running this code and it only works when the Server gives a 200 Code for the same web page.

It's a small piece of code only a few lines, so maybe worth considering?

----------

@Samuell1 Suggestion, I have seen in another issue somewhere, I have read several people asking for this request'. I agree with them and it would be a good idea to have a single css file and single js file.

I got a PR open right now found here: https://github.com/octobercms/october/pull/4285

But it would be nicer to merge the 'extra' files into a single file for both css and js.

----------

@LukeTowers I totally agree with you, when I wrote down some of these ideas, I feel they are not low hanging fruit and would require quite a lot of coding. As you labelled this as a '**discussion**' I thought I'd add them here anyway.

Totally with you on creating smaller issues and pr's from this discussion.

Also the dashboard loading is terrible. I have quite a few plugins for the dashboard area but two I want to talk about are:

- RainLab's Google Analytics Plugin
- Scott Bedard's Google Analytics Extension Plugin

I spoke to Daftpunk a few years ago about this. Just to be clear the issue is not from his RainLab plugin. It's from Scott Bedard's plugin. The RainLab plugin is required to make Scott's plugin work.

The issue there is basically it takes between 5 - 10 seconds for the dashboard to load all the data from Google Analytics.

The Plugin stops the dashboard from loading straight away and that maybe a fix we could do.

Making the backend load straight away and not get slowed down by third-party plugins.

----------

Any issues that are good people can comment and give +1 and then can open another issue for them.

If can keep this issue open for a while then more people maybe can add their input and I be up for helping improve performance issues with October.

p.s. I got a dozen PR's open right now and sitting under testing. So I'm just waiting for them to get processed before starting new things. But always happy to help October Team.

**Comment by LukeTowers** [comment]
I feel like dashboard performance could be improved by implementing caching of the widgets contents with perhaps an in widget refresh button to clear the cache for that specific widget instance.

**Comment by ayumihamsaki** [comment]
I was thinking loading the dashboard core widgets first and then loading third party widgets afterwards. First by loading the widget code and then filling the content in. Also preloading the core code minus the content.

I draw what I mean maybe easier to understand me:

![Untitled](https://user-images.githubusercontent.com/17784082/59365032-bf103300-8d2f-11e9-8dd3-299a4a56ea20.png)

**Comment by LukeTowers** [comment]
I'm not convinced that's the best way of doing things, seems needlessly complex

**Comment by ayumihamsaki** [comment]
Fair enough, my idea is based on 'trust models' we use in coding nowadays. We assume the core code is clean and the majority of errors would lie in the third party plugins that would cause issues to the loading.

**Comment by ayumihamsaki** [comment]
@LukeTowers Is it ok and are you happy to open an issue up with regards to Samuell1 point about

To Minify the assets:

- framework.combined-min.js
- framework.extras-min.css

In production enviroments.

**Comment by ayumihamsaki** [comment]
I've just seen some small issues with the code, where we could probably fix:

The backend signin screen has two files:

`modules/backend/assets/js/auth/auth.js?v=456`

with the code:

```javascript
$(document).ready(function(){
    $(document.body).removeClass('preload')

    $('form input[type=text], form input[type=password]').first().focus()
})
```

and the file

`modules/backend/assets/js/auth/uninstall-sw.js`

with the code:

```javascript
// Only run on HTTPS connections
if (location.protocol === 'https:') {
    // Unregister all service workers before signing in to prevent cache issues
    navigator.serviceWorker.getRegistrations().then(
        function(registrations) {
            for (let registration of registrations) {  
                registration.unregister();
            }
    });
}
```

These are really small pieces of code and instead of wasting two look-ups you could just inline this javascript in the head section to speed up things!

Another issue I have just seen is that the uninstall service worker script is only running on the signup webpage and not throughtout the backend. Which is letting my Service Worker from the front-end run in the backend. Which is caching files in the backend, but more worrying I could inject code into the Service Worker from the front-end to run in the backend.

So I would like to also block off the Service Worker from the front-end running in the backend.

**Comment by bennothommo** [comment]
@ayumihamsaki 

> In my Blog Plugin I did use php as you suggested to create the 304 redirect, the code example I used was a modified version found here: https://www.php.net/manual/en/function.header.php

Nice! Yeah, we can probably do 304 Not Modified calls for pages themselves (ie. CMS, Static Pages, Blogs etc.) which have to run through the PHP pipeline anyway, but my comment was referring to other assets like CSS and JS.

**Comment by ayumihamsaki** [comment]
@bennothommo Cool, I get what you are saying now! Sorry I misunderstood the first time around what you meant.

Yes please can we have 304 on the pages and not on the css and js.

**Comment by ayumihamsaki** [comment]
Going to show you a screenshot of the backend Dashboard, it shows the dashboard having a 8 second (wait) delay in Google Chrome while the assets are loaded via JSON:

![Untitled](https://user-images.githubusercontent.com/17784082/59454711-9741cd80-8e0a-11e9-8693-c18406ef270b.png)

**Comment by ayumihamsaki** [comment]
Q: Can we do something about storm-min.js there so much garage code inside.

1. It's 188Kb in size, could we minify it please to under 100Kb.

2. We got encrypted code:

![1](https://user-images.githubusercontent.com/17784082/59775105-7a911400-92a8-11e9-80ea-0717aa946019.png)

3. We got a huge amount of strange Hexadecimal Code:

![1](https://user-images.githubusercontent.com/17784082/59775205-af9d6680-92a8-11e9-9431-fa1395410bb1.png)

4. Not sure what's going on with this, if it's commented out or not etc.

![1](https://user-images.githubusercontent.com/17784082/59775268-d6f43380-92a8-11e9-851d-6a9a3c73649a.png)

I can see why it's takes so long to load and Google Canary scanning the code for any nasty things.

**Comment by ayumihamsaki** [comment]
Q: Can we do something about storm-min.js there so much garage code inside.

1. It's 188Kb in size, could we minify it please to under 100Kb.

2. We got encrypted code:

![1](https://user-images.githubusercontent.com/17784082/59775105-7a911400-92a8-11e9-80ea-0717aa946019.png)

3. We got a huge amount of strange Hexadecimal Code:

![1](https://user-images.githubusercontent.com/17784082/59775205-af9d6680-92a8-11e9-9431-fa1395410bb1.png)

4. Not sure what's going on with this, if it's commented out or not etc.

![1](https://user-images.githubusercontent.com/17784082/59775268-d6f43380-92a8-11e9-851d-6a9a3c73649a.png)

5. There are somethings ref webpack and we are not using it.

I can see why it's takes so long to load and Google Canary scanning the code for any nasty things.

**Comment by ayumihamsaki** [comment]
Q: Can we do something about storm-min.js there so much garage code inside.

1. It's 188Kb in size, could we minify it please to under 100Kb.

2. We got encrypted code:

![1](https://user-images.githubusercontent.com/17784082/59775105-7a911400-92a8-11e9-80ea-0717aa946019.png)

3. We got a huge amount of strange Hexadecimal Code:

![1](https://user-images.githubusercontent.com/17784082/59775205-af9d6680-92a8-11e9-9431-fa1395410bb1.png)

4. Not sure what's going on with this, if it's commented out or not etc.

![1](https://user-images.githubusercontent.com/17784082/59775268-d6f43380-92a8-11e9-851d-6a9a3c73649a.png)

5. There are somethings ref webpack and we are not using it.

I can see why it's takes so long to load and Google Canary scanning the code for any nasty things.

When I try to load a backend web page I really dont need 2/3 of this junk. For example, do I really need to load every single language pack at the same time?

**Comment by bennothommo** [comment]
@ayumihamsaki `storm-min.js` is the compilation of several libraries into one - [see here](https://github.com/octobercms/october/blob/master/modules/system/assets/ui/storm.js) for the list. I can say that (3) is likely moment.js and (4) are templates for certain widgets.

I'm sure there are ways to abstract certain libraries out of there and to load them only when needed, but truth be told, ~180kb of JS file for a whole heap of backend functionality, I personally feel, is not that big a deal.

**Comment by ayumihamsaki** [comment]
Also would it be a good idea for October to start using some code size limits and create some standards for everything, for example something like this:

![Untitled](https://user-images.githubusercontent.com/17784082/59776560-f724f200-92aa-11e9-8fea-3d3a6c6890c6.png)

**Comment by ayumihamsaki** [comment]
Thanks @bennothommo for comment.

> I'm sure there are ways to abstract certain libraries out of there and to load them only when needed, but truth be told, ~180kb of JS file for a whole heap of backend functionality, I personally feel, is not that big a deal.

It's a big deal for me using Google Canary as it's all getting virus scanned under Google Chrome's new Manifest V3 spec.

**Comment by ayumihamsaki** [comment]
@bennothommo Thanks for the link. Can't we split `storm.min.js` into two separate files.

1. Load the essential files in the `<head>`.
2. Load the extra files in an `async` near the `</body>` closing tag.

**Comment by LukeTowers** [comment]
@ayumihamsaki you're welcome to try to split it and see how the performance is affected. Performance improvements on hard data only please 😉

**Comment by ayumihamsaki** [comment]
@LukeTowers Thanks I will try to do, my theory is like this:

p.s. 

> I'm sure there are ways to abstract certain libraries out of there and to load them only when needed, but truth be told, ~180kb of JS file for a whole heap of backend functionality, I personally feel, is not that big a deal.

That would be fine if we were all running `HTTP 1.1` but with `HTTP/2` it makes no sense to try and pipe down a massive file.

On a timing diagram it would look like this:

![Untitled](https://user-images.githubusercontent.com/17784082/59779080-859b7280-92af-11e9-8c68-3d819fe291ff.png)

Hopefully I can prove it with some results.

**Comment by bennothommo** [comment]
@ayumihamsaki I agree with you, but from a pragmatic standpoint, we cannot assume everyone running October is running a HTTP/2 compatible server.

**Comment by ayumihamsaki** [comment]
@bennothommo totally agree with you.

**Comment by ayumihamsaki** [comment]
Just had an idea! 💡 

We have all talked about to minify or not to minify the backend assets many times.

Why don't we just create both sets and then attach them to the config/environment.php

See this code line here:

https://github.com/octobercms/october/blob/7590dfd30fb08242c6c8e111115659d38ac4aeb8/config/environment.php#L16

## Development Example:

`'default' => 'development',`

Assets are set to un-minify.

## Production Example:

`'default' => 'production',`

Assets are set to minify.

Then all I need to do is create a wrapper around the `storm` and` october` `js` and `css` files.

**Comment by ayumihamsaki** [comment]
Just had an idea! 💡 

We have all talked about to minify or not to minify the backend assets many times.

Why don't we just create both sets and then attach them to the config/environment.php

See this code line here:

https://github.com/octobercms/october/blob/7590dfd30fb08242c6c8e111115659d38ac4aeb8/config/environment.php#L16

## Development Example:

`'default' => 'development',`

Assets are set to un-minify.

## Production Example:

`'default' => 'production',`

Assets are set to minify.

Then all I need to do is create a wrapper around the `storm` and` october` `js` and `css` files.

Or we can create another config variable.

Would ask Luke to decide where to create a new variable in the config area, as I can't decide between: `config/cms.php` or `config/environment.php` or `config/testing.php`

Also the CMS could check the deployment of where it is located. I notice a lot of reviewers are using `localhost` to test things, so maybe do the following:

## Localhost

Load the following assets:

- october.css
- storm.css
- october.js
- storm.js

This saves time for the testers to have to set an option in the config settings area.

## Config (Development) - Hosted on a Live Server

Load the following assets:

- october.css
- storm.css
- october.js
- storm.js

## Config (Production) - Hosted on a Live Server

Load the following assets:

- october.min.css
- storm.min.css
- october.min.js
- storm.min.js

Hopefully this then allows some performance enhancements to people not writing and testing pull requests for October Source Code.

Default setting would be: `Config (Production)`

**Comment by LukeTowers** [comment]
If we went that way @ayumihamsaki then we'd use app.debug to determine to load minified or unminified. Again, hard numbers only 😉

**Comment by ayumihamsaki** [comment]
ok 😄

**Comment by ayumihamsaki** [comment]
## Notes

Just gonna add some notes, two PR I want to start with:

1. Fix the top menu navigation (it has really bad render blocking). As per my comment here: https://github.com/octobercms/october/issues/2820#issuecomment-511031119

2. Split the main assets into sync and async.

Hopefully can get these two things coded in the near future - but now on my to do list.

**Comment by ayumihamsaki** [comment]
## Notes - To do list (Performance)

Just gonna add some notes, three PR I want to start with:

1. Fix the top menu navigation (it has really bad render blocking). As per my comment here: https://github.com/octobercms/october/issues/2820#issuecomment-511031119

2. Split the main assets into sync and async.

3. Rip out Isotope and replace it with a more modern drag n drop, as per comment: https://github.com/octobercms/october/issues/3958#issuecomment-508539260

Hopefully can get these three things coded in the near future - but now on my to do list.

**Comment by Samuell1** [comment]
I findout that Yaml parsing is not cached and it takes a long time. It should make backend a little bit faster. https://github.com/octobercms/october/pull/4526

**Comment by ayumihamsaki** [comment]
Lazyload top navigation icons and remove external loading of default avatar https://github.com/octobercms/october/pull/4562

**Comment by LukeTowers** [comment]
@ayumihamasaki2019
